### DISCUSSION TITLE
Authentication using ClientID and Secret

### DISCUSSION THREAD

**Discussion by gavdgavd** [pnp/pnpcore#738] [Q&A]
Looking for help about the use of authentication in PnP Core SDK using the AD Client ID with a Secret.
Working to get data from SharePoint using the PnP Core SDK, I must use an Azure App Registration with Application rights (not Delegated rights); thus, I do have only the ClientID and the Secret. 
The only way to authenticate using the PnP Core SDK with a secret seems to be using OnBehalfOfAuthenticationProvider. Unfortunately, the fourth parameter of OnBehalfOfAuthenticationProvider is a function that provides the OAuth bearer Token from Azure (the first three parameters are known: ClintId, TenantId, and Secret). But at the moment of creation of the Builder, the call hasn't made yet a trip to Azure, thus the Token is unknown. Additionally, I do not see the reason why the Token must be given as input parameter to ask for the Token again.
Tried to use ExternalAuthenticationProvider as well, giving the Token obtained through a direct call to Azure, but that is very convoluted, and I am getting the cryptic error "One or more errors occurred".
Any example and/or idea will be very appreciated. Thanks in advance.

**Comment by jansenbe** [comment]
Hi @gavdgavd ,

App-only authentication against SharePoint Online requires certificate based authentication for calling the "classic" SharePoint REST/CSOM APIs. The SharePoint Graph calls can work with clientid+secret, but since PnP Core SDK requires both type of APIs (as not all features are exposed via the Graph APIs) you need to use certificate based auth. See https://pnp.github.io/pnpcore/demos/README.html for our samples, the **Authentication Type** column in the tables shows the used auth provider and allows you to pick a relevant sample.

**Reply by gavdPublic** [reply]
Hi @jansenbe,
Thanks for the information. Please allow me to set it in my own words to see if I understand it correctly:
- Using the PnP Core SDK to manage information in SharePoint
- I CAN create a PnP.Core.Services.PnPContext through the PnP.Core.Services.Builder.Configuration.PnPCoreOptions.DefaultAuthenticationProvider using Account+PW (UsernamePasswordAuthenticationProvider), a Certificate (X509CertificateAuthenticationProvider), or Interactively (InteractiveAuthenticationProvider).
- I CANNOT create a PnP.Core.Services.PnPContext through the authentication providers van PnP Core SDK using a Secret; I must use a Certificate. (The question here is what happens if I MUST use a Secret and cannot use a Certificate?)
Please correct me if I am wrong.
Many thanks in advance. Gr. Gustavo

**Reply by jansenbe** [reply]
@gavdgavd : that's correct, for delegated permissions you use username/pwd or interactive or onbehalfoff, for application permissions you can only use certificates. SharePoint APIs require certificates for app-only, the Graph APIs also work fine with certificates so there shouldn't be a scenario that mandates authentication via secret.

**Reply by gavdgavd** [reply]
Hi again @jansenbe. Thanks for the confirmation. I would like to gently disagree with your assessment that there shouldn't be a scenario that mandates authentication via secret.
Currently we are working on a project by a client that (by internal policies) forces us to use only and exclusively Azure App Registrations with Secrets. They have a fully automated process where Secrets are created on-the-fly when an application tries to use the App Registration; there is a flow behind that requires authorization to create the Secret, and the Secret is deleted when the app has finished working. In this way, each application can use a secret only one time, and the company maintains full control over when an application is run, by who, and who authorized it.
At this moment, we can use Graph getting a Graph-Client (Microsoft.Graph.GraphServiceClient) through the provider Microsoft.Graph.Auth.ClientCredentialProvider that gets the Azure Token using a Microsoft.Identity.Client.IConfidentialClientApplication that query AD using the App Registration Secret. Once we have the Graph-Client, we can get everything from Graph using the Microsoft.Graph namespace, including retrieving information saved in SharePoint.
Investigating the possibility of using the PnP Core SDK to replace the calls made using Microsoft.Graph and Microsoft.Graph.Auth, we struggled to authenticate because we don't have a provider to work with Secrets. Again, we MUST use Secrets, there is no option to use Certificates because the enterprise architecture inside the company enforces the use of Secrets.
Any idea how we can use the PnP Core SDK in this scenario will be appreciated. Thanks in advance. Gustavo

**Reply by jansenbe** [reply]
@gavdgavd : that's and interesting use case you have :-) I would love to offer support for clientid+secret for PnP Core SDK, but SharePoint APIs (REST/CSOM) simply do not work when one uses clientid+secret. This is not a library limitation, it's how things are built by the SharePoint team. So as long as the SharePoint does not support clientid+secret PnP Core SDK can't support it either. 

This means in your case that you'll have to limit yourselves to the SharePoint Graph APIs and not use PnP Core SDK.

**Reply by gavdgavd** [reply]
Hi @jansenbe. Thank you for the explanation. And, indeed, the security schema is remarkably interesting and original. I need to say, I love it, although it makes our life sometimes a little bit harsh. The environment of the customer is highly (very, very highly) secured, touching paranoia, but also very safe. Our problem with the Microsoft.Graph.Auth that we currently use is that the library has been deprecated by Microsoft, and we need to review all the code now. Recommended by Microsoft is to use the Azure.Identity namespace, but we have problems implementing it. That is the reason we are investigating the PnP Core SDK.
Anyway, thank you very much for the explanation, it is pretty clear. You and the PnP team are doing an impressive job with the SDK, thanks for that as well.
Gr., Gustavo.
PS: I am wondering if it is possible to have a provider that accepts the bearer token as input parameter to create the DefaultAuthenticationProvider and the PnPContext. Something like the ExternalAuthenticationProvider but accepting the full token instead of a URL.

**Reply by jansenbe** [reply]
Hi @gavdgavd , our `ExternalAuthenticationProvider` implementation does require you to provide us the access token: we give you the needed resource uri and scopes and you return the access token (see https://github.com/pnp/pnpcore/blob/dev/samples/Demo.ASPNetCore/Controllers/HomeController.cs#L138 for a sample). But this will still fail when you use clientid+secret because SharePoint will be check from the token how is was created and it will see this token was not created by using a certificate and therefore throw an error.

**Reply by aathar** [reply]
hi @jansenbe  we have a similar requirement. We were able to get the clientId and clientSecret using the sharepoint. What we are trying to achieve is to be able to upload files to Sharepoint from our public facing website. The website sends the file to the backend api (C#) and the controller action would authenticate using clientId and clientSecret, and then once authenticated, will upload the document in sharepoint.

We were able to get the authentication by using 
 `var clientContext = new AuthenticationManager().GetACSAppOnlyContext(siteUrl, clientId, clientSecret);`
          
but that is pnp.Framework. We want to be able to use pnpCore as the documentation to upload large files look promising. But when we try to use pnpCore, during authentication "integrated", it takes us to the azure login page. We can't let our public users go to azure login page, that should take place in the backend. Can you guide us to the right direction please?

**Reply by jansenbe** [reply]
Hi @aathar ,

Using Azure ACS is not natively possible in PnP Core SDK as Azure ACS is a legacy authentication concept that for the most part already has been deprecated. In SharePoint it's still supported for the moment. What you should do is use certificate based authentication and setup this up in Azure AD (see https://pnp.github.io/pnpcore/demos/Demo.AzureFunction.OutOfProcess.AppOnly/readme.html for a sample that uses certificate based auth and which shows how to set this up).

**Reply by gavdgavd** [reply]
Hi @jansenbe Bert,
I know we discussed this sometime ago, but the situation is still bottering me, probably you can enlighten me.

Microsoft deprecated the Azure ACS for SharePoint, that is known. As it is also known that ACS "a kind of" secret based authentication was, and that it is still useable for SharePoint Add-Ins, but not for normal calls against the server. But an Azure App Registration with a Secrets schema is a different thing than the ACS.

I can send a query to Microsoft Graph to get information from SharePoint using an Azure App Registration based on Secrets authentication. It just works fine; I can give you the code if you are interested (PowerShell using the normal "Invoke-RestMethod" and "Invoke-WebRequest" cmdlets). I don't think Graph is using the Azure ACS, but, of course, I am not sure about that because I do not know how Graph works internally.

Do you have any idea why I can use Graph in this way to approach SharePoint, but it doesn't work using other methods?

I appreciate your comments, thanks a lot.
Gr.,
Gustavo

**Reply by jansenbe** [reply]
@gavdgavd : SharePoint REST/CSOM APIs only work with certificate auth when using an Azure AD app setup, pure Graph calls can work with clientid/secret, but since PnP Core SDK requires SharePoint REST we don't support that option. We'll not add ACS support due to above explained reasons.

**Reply by yueming-zhang** [reply]
@jansenbe, I found your reply is very insightful, Hope you can help. 

I need to access SharePoint online from Linux OS. Could you point me out what's the best way to generate certificate on Ubuntu? Most of the online helps today are using PowerShell. 

Alternatively, if I generate the cert using PowerShell on a windows OS, and then move it Ubuntu, I don't think it will work. But Please correct me. 

Thank you in advance.

**Reply by jansenbe** [reply]
@yueming-zhang : see https://learn.microsoft.com/en-us/dotnet/core/additional-tools/self-signed-certificates-guide#with-openssl

**Reply by ParvathySN** [reply]
> @gavdgavd : SharePoint REST/CSOM APIs only work with certificate auth when using an Azure AD app setup, pure Graph calls can work with clientid/secret, but since PnP Core SDK requires SharePoint REST we don't support that option. We'll not add ACS support due to above explained reasons.

@jansenbe So For SharePoint CSOM APIs, we have to use certificate?

**Reply by bcameron1231** [reply]
> > @gavdgavd : SharePoint REST/CSOM APIs only work with certificate auth when using an Azure AD app setup, pure Graph calls can work with clientid/secret, but since PnP Core SDK requires SharePoint REST we don't support that option. We'll not add ACS support due to above explained reasons.
> 
> @jansenbe So For SharePoint CSOM APIs, we have to use certificate?

Yes. Any call into SharePoint APIs with an app require a certificate.
